VSR: Refactor+fix DVC/SV header generation

sentientwaffle opened the pull request:
## Refactor

Refactor DVC/SV header generation for greater clarity (at the cost of increased verbosity). Specifically, the critical logic for each different case (retired primary/follower, durable SV/DVC, etc) are made as clear as possible. The split cases also enable stricter assertions of chain states.

This new implementation also includes additional headers during the "suffix_done" phase by skipping over "same-view gaps". (That is, given `1,_,3`, if op=1 and op=3 have the same view, then if op=3 is included in the DVC then op=1 is guaranteed to be part of the same hash-chain, so it is safe to include as well).

## Fix

Cases with a durable SV now properly use that durable SV's head op in place of the DVC anchor.

## Pre-merge checklist

Performance:

* I am very sure this PR could not affect performance.

sentientwaffle committed:
VSR: Don't send SVC on recovering_head→SV

When a replica receives a SV, it uses `view_jump()`, which calls `transition_to_view_change_status()`.
(The key part of this change is at the end of `transition_to_view_change_status()` — the rest is just refactoring.)

sentientwaffle committed:
VSR: Refactor+fix DVC/SV header generation

Refactor DVC/SV header generation for greater clarity (at the cost of increased verbosity).
Specifically, the critical logic for each different case (retired primary/follower, durable SV/DVC, etc) are made as clear as possible.
The split cases also enable stricter assertions of chain states.

This new implementation also includes additional headers during the "suffix_done" phase by skipping over "same-view gaps".
(That is, given `1,_,3`, if op=1 and op=3 have the same view, then if op=3 is included in the DVC then op=1 is guaranteed to be part of the same hash-chain, so it is safe to include as well).

Fix: Cases with a durable SV now properly use that durable SV's head op in place of the DVC anchor.

sentientwaffle committed:
VSR: Trigger start_view after view_durable_update

Fixes a regression from #426.
Prior to #426 we would send a `start_view` as part of starting the new primary.
That was removed since the replica is updating its `view_durable` at the end of `primary_start_view_as_the_new_primary()`, so it would be dropped there anyway (`view > view_durable`).

I (mistakenly) thought that the `start_view` would trigger via a timeout, but that is not the case.

Unrelated: Add a couple timeout resets — these aren't necessary, but can help avoid unnecessary view-changes if the (superblock) disk latency is high.

sentientwaffle committed:
VSR: Don't send start_view until log_view durable

Don't send a `start_view` until the corresponding `log_view` is durable.
(We already waited for the `view` to be durable — but it always is, since we would have persisted our DVC before participating in the view-change).
Surprisingly, this wasn't actually detected by the VOPR — possibly because network faults actually _decrease_ the likelihood of triggering this bug.

What we want to avoid is:
1. Participate in view-change as new primary. (DVC is durable).
2. Receive DVC quorum. Start persisting the SV.
3. Send SV.
4. Other replicas receive SV.
5. Prepare+commit some messages.
6. Crash (SV was never persisted — we backtrack to sending the DVC from step 1, and forget all of the prepares (and maybe even commits) in the latest view).

matklad commented on src/vsr.zig:
```diff
@@ -1007,26 +1007,30 @@ pub fn quorums(replica_count: u8) struct {
-/// - DVC headers (disjoint chain)
-pub const ViewChangeHeaders = struct {
+pub const Headers = struct {
+    pub const Array = std.BoundedArray(Header, constants.pipeline_prepare_queue_max);
```
This is pre-existing, but I don't feel great about this array -- it's sizeof is about a kilobyte, and, by using an array here, we are asking for memcpys.

matklad replied:
That is, I'd expect us to write to some exiting memory, rather than return it

matklad commented on src/vsr.zig:
```diff
@@ -1098,3 +1102,219 @@ test "ViewChangeHeaders.view_for_op" {
+            },
+        },
+    ) ViewChangeHeadersArray {
+        const current = options.current;
```
large memcpy?

sentientwaffle replied:
Yeah, I'm thinking I'll change these from `BoundedArray`s to `*BoundedArray`s.
Then the physical allocation will be part of the replica, e.g.:

```
        view_change_headers: vsr.Headers.ViewChangeArray = .{ .buffer = undefined },
        journal_headers: vsr.Headers.Array = .{ .buffer = undefined },
```

which is a little awkward (they aren't really state, just scratch space for `create_view_change_headers()` to put things).

matklad replied:
Ah, I think, because it is const, zig should be smart enough to treat this as an alias, right?

sentientwaffle replied:
I think it should be, but isn't. In this particular case it could be ` = &options.current`, though.

sentientwaffle replied:
But that still involves a memcopy when the function is called.

sentientwaffle replied:
https://github.com/tigerbeetledb/tigerbeetle/pull/447/commits/2e8be5452d1fa4cdbf923547d83fc5f27350fb38

matklad commented on src/vsr.zig:
```diff
@@ -1098,3 +1102,219 @@ test "ViewChangeHeaders.view_for_op" {
+        return .{ .array = array };
+    }
+
+    pub fn build(
```
Might make sense to point to comment on `DVCQuorum`, as this uses terminology we introduce there (gap, break)

sentientwaffle committed:
VSR: Avoid memcpy of view-change headers

matklad commented on src/vsr.zig:
```diff
@@ -1098,3 +1102,221 @@ test "ViewChangeHeaders.view_for_op" {
+                }
+            };
+
+            if (current.log_view == current.view) {
```
Maybee `command_current == .start_view` to deduplicate the condition?

Would also appreciate a comment on `const command_current:` explaining the semantics. I understand what `current.log_view == current.view` condition means (the current view is an active one), but I don't understand why that translates to the two commands. 

Maybe its better to pass command in, like we do in `create_view_change_message`, and assert?

Or even, pass the command in for _current_, assert, and then use asserted logic to re-derive it for durable?

sentientwaffle replied:
> Maybee `command_current == .start_view` to deduplicate the condition?

:+1: 

> Would also appreciate a comment on `const command_current:` explaining the semantics. I understand what `current.log_view == current.view` condition means (the current view is an active one), but I don't understand why that translates to the two commands.

Does this help?

```zig
        // The rules for generating DVCs and SVs differ. We use the current view numbers to
        // determine which is being generated:
        // - when `log_view < view`, generate a DVC
        // - when `log_view = view`, generate a SV
        const command_current: enum { start_view, do_view_change } =
            if (current.log_view == current.view) .start_view else .do_view_change;
        // Likewise, the durable view numbers identify whether the durable headers were from a past
        // DVC or SV. The durable headers are only useful if they are from the same view as our
        // current headers, though.
        const command_durable: enum { start_view, do_view_change, outdated } = command: {
            if (durable.log_view == current.log_view) {
                if (durable.log_view == durable.view) {
                    break :command .start_view;
                } else {
                    break :command .do_view_change;
                }
            } else {
                break :command .outdated;
            }
        };
```

> Maybe its better to pass command in, like we do in create_view_change_message, and assert?
> Or even, pass the command in for current, assert, and then use asserted logic to re-derive it for durable?

In `create_view_change_message` we assert:

```
            assert(command != .do_view_change or self.log_view < self.view);
            assert(command != .start_view or self.log_view == self.view);
```

Passing in the command there isn't strictly necessary — it can be derived entirely from the replica state. But passing in the command there enables a little extra verification of the call site.


I don't pass the command to `build()` because:
- We would have the `command` via the `create_view_change_message` call site, but the `view_durable_update` call site would have to re-derive it anyway, using the exact same method.
- (Less important:) I like that the `current`/`durable` parameters mirror each other. And `view`/`log_view` provides more information than the `command` alone.

matklad replied:
>Likewise, the durable view numbers identify whether the durable headers were from a past  DVC or SV.

This is still confusing to me, what polarity are we talking about? _sending_ DVC or SV (where we create VC message to send) or _receiving_ DVC or SV (where we replace our headers using received message[s], and then create VC message for `view_durable_update`)

sentientwaffle replied:
It is kind of both, but also they are sort of the same thing :sweat_smile: 

This function is used for all of the following:
- primary generating a SV to send to its backups
- primary generating a SV to write to its superblock
- primary/backup generating a DVC to send
- primary/backup generating a DVC to write to its superblock
- And this is a weird one: backups generating a SV which they write to their superblock. This is mostly for symmetry, since of course a backup will never send a SV message over the network. But since we only ever do this after receiving (and installing) a SV from the primary, it follows the same rules as the primary's SV generation.

sentientwaffle committed:
VSR: Document header commands

sentientwaffle committed:
VSR: DVC/SV header documentation

sentientwaffle commented:
Changes:
- Rebased on main.
- Added documentation.
- Cherry-picked a few start-view fixes from another branch I'm working on.

sentientwaffle committed:
VSR: Rearrange view_durable_update

sentientwaffle committed:
VSR: Use correct Array capacity

matklad commented on src/vsr/replica.zig:
```diff
@@ -4838,19 +4737,23 @@ pub fn ReplicaType(
+                if (message.header.command == .do_view_change or
+                    message.header.command == .start_view)
+                {
+                    if (self.log_view_durable() < self.log_view) {
```
I'd maybe use `message.header.timestamp` rather than `self.log_view` here. THey are he same, `timestamp` is a bit more obscure, _but_ it emphasizesthat we care about data in the message. Though, both are weird tbh, no strong opinion here.

matklad replied:
Ah, perhaps `const message_log_view = message.header.timestamp;` is the best way to spell this.

sentientwaffle replied:
I slightly prefer how it is right now, since `message.header.timestamp` is only set for `do_view_change`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -4749,7 +4648,7 @@ pub fn ReplicaType(
                     assert(message.header.replica == self.replica);
                     assert(message.header.op == self.op);
                     assert(message.header.commit == self.commit_min);
-                    assert(replica == self.primary_index(self.view));
```
Why do we delete this assert? It should still hold, no?

sentientwaffle replied:
Ah that snuck in via `git add --patch`. I'm ok leaving it removed b/c it will very shortly not be true, since DVCs will be broadcast to all replicas.

matklad commented on src/vsr.zig:
```diff
@@ -1098,3 +1102,251 @@ test "ViewChangeHeaders.view_for_op" {
+
+            if (command_current == .start_view) {
+                // Primary: Collect headers for a start_view message.
+                // Follower: these headers are stored in the superblock's vsr_headers.
```
```suggestion
                // Backup: these headers are stored in the superblock's vsr_headers.
```

matklad commented on src/vsr.zig:
```diff
@@ -1098,3 +1102,251 @@ test "ViewChangeHeaders.view_for_op" {
+                    // the durable SV to prune bad headers by their view numbers.
+                    .chain_break => unreachable,
+                }
+                suffix_done = op <= op_head_durable;
```
When we use dvc_anchor for suffix, we cap it at `std.math.max(options.op_checkpoint, ...`, presumably because before checkpoint anything goes. Shouldn't be use the same capping for `op_head_durable` as well?

sentientwaffle replied:
I'm not 100% sure but I think it is correct as-is. This branch (and the other `op <= op_head_durable`) are both for `command_durable == .start_view`. So:
- We are generating a DVC.
- We have a durable SV.
- That durable SV's `log_view` is the same as the current `log_view` (i.e. the `log_view` of the DVC).

 https://github.com/tigerbeetledb/tigerbeetle/blob/811740b61f82b5b973cbbb90b56bb58eb7af2511/src/vsr/journal.zig#L1189-L1211 (part of journal recovery) helps by cleaning up the log. And that covers all slots, using the `view` number, so it doesn't care about the checkpoint at all.

sentientwaffle replied:
And if we _didn't_ just recover, then installing the `start_view` would have truncated everything after the `start_view`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -4993,13 +4896,27 @@ pub fn ReplicaType(
+                } else {
+                    self.send_prepare_oks_after_view_change();
+                }
+            }
```
Shouldn't we also `self.send_do_view_change()` here `if self.start_view_change_quorum` ?

sentientwaffle committed:
VSR: Send DVC after view_durable_update

This isn't required since DVCs are retried on a timeout interval, but it is good to send the first DVC promptly in the regular case.

sentientwaffle committed:
VSR: Update docs

matklad commented on src/vsr/replica.zig:
```diff
@@ -4917,6 +4919,14 @@ pub fn ReplicaType(
+
+            if (self.status == .view_change and self.log_view < self.view) {
+                if (self.primary_index(self.view) != self.replica and
+                    self.start_view_change_quorum and !self.do_view_change_quorum)
```
I am not entirely sure about this last `!self.do_view_change_quorum` condition, but I guess our plan to broadcast DVCs woudl change the calculus here anyway

sentientwaffle replied:
Ah yeah my brain has already switched over to the new DVCs. It isn't necessary with the old DVC, but isn't incorrect either.

matklad commented:
bors r+
